pkg/gatewayserver/io/basicstationlns/messages/downstream.go

```diff
@@ -0,0 +1,117 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package messages
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"time"
+
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+)
+
+var errDownlinkMessage = errors.Define("downlink_message", "could not translate downlink message")
+
+// DownlinkMessage is the LoRaWAN downlink message sent to the basic station.
+type DownlinkMessage struct {
+	DevEUI      string  `json:"DevEui"`
+	DeviceClass uint    `json:"dC"`
+	Diid        int64   `json:"diid"`
+	Pdu         string  `json:"pdu"`
+	RxDelay     int     `json:"RxDelay"`
+	Rx1DR       int     `json:"RX1DR"`
+	Rx1Freq     int     `json:"RX1Freq"`
+	Priority    int     `json:"priority"`
+	XTime       int64   `json:"xtime"`
+	GpsTime     int64   `json:"gpstime"`
+	RCtx        int64   `json:"rctx"`
+	MuxTime     float64 `json:"MuxTime"`
+}
+
+// MarshalJSON implements json.Marshaler.
+func (dnmsg DownlinkMessage) MarshalJSON() ([]byte, error) {
+	type Alias DownlinkMessage
+	return json.Marshal(struct {
+		Type string `json:"msgtype"`
+		Alias
+	}{
+		Type:  TypeDownstreamDownlinkMessage,
+		Alias: Alias(dnmsg),
+	})
+}
+
+// FromDownlinkMessage translates the ttnpb.DownlinkMessage to LNS DownlinkMessage "dnmsg".
+func FromDownlinkMessage(ids ttnpb.GatewayIdentifiers, down ttnpb.DownlinkMessage, dlToken int64, dlTime time.Time, latestXTime int64) DownlinkMessage {
+	var dnmsg DownlinkMessage
+	scheduledMsg := down.GetScheduled()
+	dnmsg.Pdu = hex.EncodeToString(down.GetRawPayload())
+	dnmsg.RCtx = int64(scheduledMsg.Downlink.AntennaIndex)
+	dnmsg.Diid = dlToken
+
+	// This field is not used but needs to be defined for the station to parse the json.
+	dnmsg.DevEUI = "00-00-00-00-00-00-00-00"
+
+	// Chosen fixed values.
+	dnmsg.Priority = 25
+	dnmsg.RxDelay = 1
+
+	// Fix the Tx Parameters since we don't use the gateway scheduler.
+	dnmsg.Rx1DR = int(scheduledMsg.DataRateIndex)
+	dnmsg.Rx1Freq = int(scheduledMsg.Frequency)
+
+	//Add the MuxTime for RTT measurement
+	dnmsg.MuxTime = float64(dlTime.Unix()) + float64(dlTime.Nanosecond())/(1e9)
+
+	// Always use the Basic Station ClassB mode for absolute time scheduling.
+	if scheduledMsg.Time != nil {
+		dnmsg.DeviceClass = uint(ttnpb.CLASS_B)
+		dnmsg.GpsTime = scheduledMsg.Time.Unix()
+		return dnmsg
+	}
+
+	dnmsg.DeviceClass = uint(ttnpb.CLASS_A)
+	// Estimate the xtime based on the timestamp; xtime = timestamp - (rxdelay+1).
+
+	t := time.Unix(int64(scheduledMsg.Timestamp), 0)
+	offset := time.Duration((dnmsg.RxDelay)) * time.Second * (1e6)
+	dnmsg.XTime = latestXTime&0x7FFFFFFF00000000 | (t.Add(-offset).Unix())
```

Comment:
Remove last parentheses

Reply:
Also we're not accounting for rollover here. This may make `dnmsg.XTime` smaller than `latestXTime`